test/yagna/helpers/negotiation.py

```diff
@@ -0,0 +1,112 @@
+"""Helper functions for building custom Offers and negotiating Agreements."""
+
+import logging
+from typing import List, Optional, Callable, Tuple, Any
+from datetime import datetime, timedelta
+
+from ya_market import Demand, DemandOfferBase, Proposal
+
+from goth.runner.probe import RequestorProbe
+from goth.runner.provider import ProviderProbeWithLogSteps
+from goth.runner.requestor import RequestorProbeWithApiSteps
+
+
+logger = logging.getLogger(__name__)
+
+
+class DemandBuilder:
+    """Helper for building custom Demands.
+
+    Use if RequestorProbeWithApiSteps.subscribe_template_demand function
+    is not enough for you.
+    """
+
+    def __init__(self, requestor: RequestorProbe):
+        self._requestor = requestor
+        self._properties = dict()
+        self._constraints = "()"
+
+    def props_from_template(self, task_package: str) -> "DemandBuilder":
+        """Build default properties."""
+
+        new_props = {
+            "golem.node.id.name": "test-requestor-{}".format(self._requestor.name),
+            "golem.srv.comp.expiration": int(
+                (datetime.now() + timedelta(minutes=10)).timestamp() * 1000
+            ),
+            "golem.srv.comp.task_package": task_package,
+        }
+        self._properties.update(new_props)
+        return self
+
+    def property(self, key: str, value: Any) -> "DemandBuilder":
+        """Add property."""
+        self._properties[key] = value
+        return self
+
+    # TODO: Building constraints.
+    def constraints(self, constraints: str) -> "DemandBuilder":
+        """Add constraints.
+
+        Note: This will override previous constraints.
+        """
+
+        self._constraints = constraints
+        return self
+
+    def build(self) -> DemandOfferBase:
+        """Create Demand from supplied parameters."""
+        return DemandOfferBase(
+            properties=self._properties,
+            constraints=self._constraints,
+        )
```

Comment:
Hmm, method chaining, as encouraged by `DemandBuilder`, is not idiomatic Python. A common convention is that methods that modify the underlying object should not return any values (other than `None`).

But I think it's fine as a helper in `test`. So it's just a side note...

Reply:
We can change this at some point, if you think, keeping python conventions is valuable.